ICRCs/ICRC-103/ICRC-103.md

```diff
@@ -0,0 +1,103 @@
+| Status |
+|:------:|
+|Draft|
+
+# ICRC-191: Enhanced Allowance Query Mechanism with Pagination
+
+## 1. Introduction
+
+Although calls to the `icrc2_approve` and `icrc2_transfer_from` methods are recorded in the ledger, it is not possible to determine the allowances that are in effect at some point in time, except by traversing the entire ledger.  This standard introduces an endpoint which will return this information thus making the management of allowances feasible.
+
+ICRC-191 is an extension of the ICRC-2 standard.  
+ICRC-191 specifies a way to list outstanding allowances.
+
+
+## 2. Metadata
+
+A ledger that implements ICRC-191 MUST include `record {name = "ICRC-191"; url = "https://github.com/dfinity/ICRC-1/standards/ICRC-191"}` as part of the output of `icrc1_supported_standards`.
+
+The endpoint introduced in this standard operates in two ways.  In the public version any principal can obtain the outstanding allowances of any other principal. In the private version, the allowances returned by the endpoint must have been issued by the caller (i.e. the caller is the principal controlling the source account of an allowance.)
+Which version of the standard is implemented by a ledger is specified through metadata which can be retrieved using `icrc1_metadata`.
+
+A ledger that implements ICRC-191 MUST return metadata `icrc191:public_allowances` of type Text (optional). The possible values are "true" if the allowance data is public and "false" otherwise.
+
+
+## 3. Methods
+
+Some of the types used are shared with standards ICRC-1 and ICRC-2; we restate their definition for completeness.
+
+```candid
+icrc191_list_allowances : (ListAllowancesArgs) -> (ListAllowancesResult) query
+
+type ListAllowancesArgs = record {
+    from_account : opt Account;
+    prev_spender : opt Account;
+    take : opt nat;
+}
+
+ListAllowanceResult = vec record {
+    from_account : Account;
+    to_spender : Account;
+    allowance : Allowance;
+}
+
+type Account = record {
+    owner : principal;
+    subaccount : opt blob;
+};
+
+type Allowance = record {
```

Comment:
I would suggest combining the record from `ListAllowanceResult` and `Allowance` along the lines of:

```candid
type Allowance = record {
  from_account : Account;
  to_spender : Account;
  allowance : nat;
  expires_at : opt nat64;
}

ListAllowanceResult = vec Allowance;
```

I would assume that the from and to are more related to the allowance than it's expiry 😅

Reply:
The type Allowance was already defined in the ICRC-2 standard and here I'm reusing. I could introduce yet another type for the flatened entries in the response (as you suggest), but is there an advantage to doing this?

Reply:
Ah that makes sense, I don't see any clear advantage for either. I suggested the flattened structure mostly to simplify the response data structure for consumers. 

My main concern would be a possible confusion due to duplicate nested `allowance` properties, where one property refers to the structure and the inner property refers to the actual amount.

Reply:
I see -- but then the same argument should apply to the Allowance type defined in ICRC-2? :)

Reply:
> I see -- but then the same argument should apply to the Allowance type defined in ICRC-2? :)

ICRC-2 doesn't have nested approvals, actually the did file doesn't even have the Approval type https://github.com/dfinity/ICRC-1/blob/main/standards/ICRC-2/ICRC-2.did

Reply:
Yes, that's definitely fair, so there's no really good reason to not flatten the entries in the list. I've implemented this in the latest version. Thanks a lot for keeping at it!